linux/daemon/OSPDaemon.c

```diff
@@ -0,0 +1,597 @@
+		return;
+	}
+	if (sz > MAX_CALSIZE) return;
+	if (sd->CalFile)
+		updateCalFile(sd->CalFile, sd->sensor[i].sensor.SensorType,
+				data, sz);
+}
+
+/*
+ * Generic alg data output call back.
+ *    Looks up the queue assocaited with the sensor and 
+ * places data on the queue.
+ */
+static void OSPDaemon_callback(ResultHandle_t handle, void *pData)
+{
+	int i;
+	struct OSPDaemon_output *out;
+
+	for (i = 0; i < sd->output_count; i++) {
+		if (handle == sd->output[i].handle)
+			break;
+	}
+	if (i == sd->output_count) return;
+	out = &sd->output[i];
+
+	OSPDaemon_queue_put(&out->q, pData);
+	for (i = 0; i < nfd; i++) {
+		if (pfd[i].fd == out->fd) {
+			pfd[i].events = POLLOUT;
+			break;
+		}
+	}
+}
+
+/* Apply modifications to the sensor data. Used with
+ * config file options to modify copied sensor data.
+ * Return: 0 - no mod, 1 - mod */
+static int modifyOutput(struct OSPDaemon_output *out, 
+		OSP_InputSensorData_t *orig,
+		OSP_InputSensorData_t *mod)
+{
+	int val[5];
+	int count;
+	int i;
+	unsigned long long ts;
+	union {
+		int i;
+		float f;
+	} conv;
+
+	if (out->format == OUT_FORMAT_INTEGER)
+		return 0;
+	count = extractOSP(out->ResultDesc.SensorType, orig, &ts, val);
+	switch (out->format) {
+	case OUT_FORMAT_FLOAT:
+		for (i = 0; i < count; i++) {
+			conv.f = val[i];
+			val[i] = conv.i;
+		}	
+		break;
+	case OUT_FORMAT_Q24:
+		for (i = 0; i < count; i++) {
+			val[i] = val[i] << 24;
+		}	
+		break;
+	case OUT_FORMAT_Q16:
+		for (i = 0; i < count; i++) {
+			val[i] = val[i] << 16;
+		}	
+		break;
+	case OUT_FORMAT_Q15:
+		for (i = 0; i < count; i++) {
+			val[i] = val[i] << 15;
+		}	
+		break;
+	case OUT_FORMAT_Q12:
+		for (i = 0; i < count; i++) {
+			val[i] = val[i] << 12;
+		}	
+		break;
+	default:
+		break;
+		
+	}
+	conv2OSP(out->ResultDesc.SensorType, mod, ts, val);
+	return 1;
+}
+/*
+ * Processes received data:
+ *    If the sensor is configured to be tee'd, queue it in the output sensor.
+ *    Send data to the alg.
+ */
+static int OSPDaemon_senddata(struct OSPDaemon_SensorDetail *s)
+{
+	OSP_InputSensorData_t *od;
+	OSP_InputSensorData_t mod;
+	OSP_STATUS_t stat;
+	int ret = 0;
+
+	if (!s->noprocess && s->pending) {
+		od = &s->pdata;
+		stat = OSP_SetInputData(s->handle, od);
+		if (stat != OSP_STATUS_OK) return 1;
+		s->pending = 0;
+	}
+
+	do {
+		od = OSPDaemon_queue_get(&s->q);
+		if (od == NULL) break;
+
+		if (s->output) {
+			if (disablepm || s->output->enable) {
+				if (modifyOutput(s->output, od, &mod) == 1) {
+					OSPDaemon_queue_put(&s->output->q, &mod);
+				} else {
+					OSPDaemon_queue_put(&s->output->q, od);
+				}
+			}
+		}
+		if (!s->noprocess) {
+			stat = OSP_SetInputData(s->handle, od);
+			if (stat != OSP_STATUS_OK) {
+				s->pending = 1;
+				memcpy(&s->pdata, od, sizeof(s->pdata));
+				ret = 1;
+				break;
+			}
+		}
+	} while(1);
+
+	return ret;
+}
+
+static int OSPDaemon_process_inbound(struct pollfd *pfd, struct pollfd *list, int nfd, int dirty)
```

Comment:
Why send both the list of pfd's and one of the values in the array? Can't we just send the index which we want to refer to?